### DISCUSSION TITLE
GitHub Copilot and PlatformIO on FreeBSD for ESP32 development

### DISCUSSION THREAD

**Discussion by probonopd** [helloSystem/hello#479] [Show and tell]
# Using GitHub Copilot and PlatformIO on FreeBSD for ESP32 development

1. Copilot refuses to run on "Code - OSS" but VSCode is not available on FreeBSD
2. PlatformIO is available for FreeBSD, but not all of the core ares, e.g., the ESP32 core isn't
3. Copilot plugin wants to store login information in "org.freedesktop.secrets" (provided GNOME Keychain) but we don't want to use anything GNOME. How can we get it to store that in a text file instead? In the very worst case, can we code our own "org.freedesktop.secrets provider" that does just enough to satisfy VSCode?
4. PlatformIO keeps saying "Initializug PlatformIO Core..." forever
5. Cannot upload

## Fix for 1

Replace `product.json` on FreeBSD with `product.json` from the original Microsoft VSCode for Linux

```
sudo mv /usr/local/share/code-oss/resources/app/product.json /usr/local/share/code-oss/resources/app/product.json.original
sudo cp VSCode-linux-x64/resources/app/product.json /usr/local/share/code-oss/resources/app/product.json
 ```

## Fix for 2

```
chmod +x ./python3.9.16-cp39-cp39-manylinux1_x86_64.AppImage
./python3.9.16-cp39-cp39-manylinux1_x86_64.AppImage --appimage-extract
mv ./squashfs-root ./python3
./python3/opt/python3.9/bin/python3.9 -m virtualenv --copies venv
# Now we have the python binaries for Linux in the env!
source ./venv/bin/activate
# Now we see (venv) in the prompt
# Now if we run vscode from there, it should hopefully use that Python?

cp -r python3/opt/python3.9/lib/python3.9/lib-dynload venv/lib/python3.9/
cp python3/usr/lib/* venv/lib

find venv/bin -exec patchelf --set-rpath '$ORIGIN/../lib' {} \;
python3
# Should now say "on linux"
```

Now always activate the venv before launching vscode

## Fix for 3

Possibly https://www.freshports.org/sysutils/pass-secrets/, but I could not find out how to use it yet.

So, https://github.com/yousefvand/secret-service to the rescue. It listens on D-Bus and saves credentials to a database.

```
git clone https://github.com/yousefvand/secret-service.git
cd secret-service
go build -race -o secretserviced cmd/app/secretserviced/main.go
go build -race -o secretservice cmd/app/secretservice/main.go
sudo mv secretservice* /usr/local/bin
sudo su
cat > /Applications/Autostart/secretservice <\EOF
#!/bin/sh

export MASTERPASSWORD=01234567890123456789012345678912
exec secretserviced
EOF
chmod +x /Applications/Autostart/secretservice
daemon /Applications/Autostart/secretservice
```

Log into GitHub Copilot. Now when restarting  Code - OSS it should automatically log in again. Only do this if no strangers have access to your computer, and if you are fine with `MASTERPASSWORD` being stored this way.

## Fix for 4

Still looking for one. As a workaround, run `pio run` from the console.

## Fix for 5

How can we get it to use FreeBSD `esptool.py`? Linux Python doesn't find the port to upload to... we need to find a way to make PlatformIO use the system-provided Python for running `esptool.py`. How?

https://github.com/platformio/platform-espressif32/issues/1037

## Questions

* Why can't the Linuxulator run the official Microsoft VSCode for Linux? There are no files ivisble in the Open/Save File dialogs

**Comment by probonopd** [comment]
* https://github.com/helloSystem/ISO/issues/479

**Comment by FabioLolix** [comment]
> Copilot refuses to run on "Code - OSS" but VSCode is not available on FreeBSD

Linked stuff is from Arch Linux's AUR, hope can you find something useful for running that on code-oss

https://aur.archlinux.org/packages/code-marketplace

https://aur.archlinux.org/packages/code-features

**Reply by probonopd** [reply]
See "Fix for 1", it worked for me.
